pkg/share/manager/cs3/cs3.go

```diff
@@ -0,0 +1,564 @@
+		return err
+	}
+
+	return m.indexer.Delete(share)
+}
+
+// ListShares returns the shares created by the user. If md is provided is not nil,
+// it returns only shares attached to the given resource.
+func (m *Manager) ListShares(ctx context.Context, filters []*collaboration.Filter) ([]*collaboration.Share, error) {
+	if err := m.initialize(); err != nil {
+		return nil, err
+	}
+
+	user, ok := ctxpkg.ContextGetUser(ctx)
+	if !ok {
+		return nil, errtypes.UserRequired("error getting user from context")
+	}
+
+	allShareIds, err := m.indexer.FindBy(&collaboration.Share{}, "OwnerId", userIDToIndex(user.GetId()))
+	if err != nil {
+		return nil, err
+	}
+	result := []*collaboration.Share{}
+	for _, id := range allShareIds {
+		s, err := m.getShareByID(ctx, id)
+		if err != nil {
+			return nil, err
+		}
+		if share.MatchesFilters(s, filters) {
+			result = append(result, s)
+		}
+	}
+	return result, nil
+}
+
+// UpdateShare updates the mode of the given share.
+func (m *Manager) UpdateShare(ctx context.Context, ref *collaboration.ShareReference, p *collaboration.SharePermissions) (*collaboration.Share, error) {
+	if err := m.initialize(); err != nil {
+		return nil, err
+	}
+	share, err := m.GetShare(ctx, ref)
+	if err != nil {
+		return nil, err
+	}
+	share.Permissions = p
+
+	data, err := json.Marshal(share)
+	if err != nil {
+		return nil, err
+	}
+
+	fn := path.Join("shares", share.Id.OpaqueId)
+	err = m.storage.SimpleUpload(ctx, fn, data)
+	if err != nil {
+		return nil, err
+	}
+
+	return share, nil
+}
+
+// ListReceivedShares returns the list of shares the user has access to.
+func (m *Manager) ListReceivedShares(ctx context.Context, filters []*collaboration.Filter) ([]*collaboration.ReceivedShare, error) {
+	if err := m.initialize(); err != nil {
+		return nil, err
+	}
+
+	user, ok := ctxpkg.ContextGetUser(ctx)
+	if !ok {
+		return nil, errtypes.UserRequired("error getting user from context")
+	}
+
+	result := []*collaboration.ReceivedShare{}
+
+	ids, err := granteeToIndex(&provider.Grantee{
+		Type: provider.GranteeType_GRANTEE_TYPE_USER,
+		Id:   &provider.Grantee_UserId{UserId: user.Id},
+	})
+	if err != nil {
+		return nil, err
+	}
+	receivedIds, err := m.indexer.FindBy(&collaboration.Share{}, "GranteeId", ids)
+	if err != nil {
+		return nil, err
+	}
+	for _, group := range user.Groups {
+		index, err := granteeToIndex(&provider.Grantee{
+			Type: provider.GranteeType_GRANTEE_TYPE_GROUP,
+			Id:   &provider.Grantee_GroupId{GroupId: &groupv1beta1.GroupId{OpaqueId: group}},
+		})
+		if err != nil {
+			return nil, err
+		}
+		groupIds, err := m.indexer.FindBy(&collaboration.Share{}, "GranteeId", index)
+		if err != nil {
+			return nil, err
+		}
+		receivedIds = append(receivedIds, groupIds...)
+	}
+
+	for _, id := range receivedIds {
+		share, err := m.getShareByID(ctx, id)
+		if err != nil {
+			return nil, err
+		}
+		metadata, err := m.downloadMetadata(ctx, share)
+		if err != nil {
+			if _, ok := err.(errtypes.NotFound); ok {
```

Comment:
Can we reverse the logic here?
```golang
if !ok {
    return nil, err
}
```
Then we can get rid of the else statement which reduces cognitive complexity.